### PR TITLE
Fix: Height for errorMessage prop on Input

### PR DESCRIPTION
**Pull request by saxenanihal95**
According to the #1755 the component showing error message should always render.

### Before
bad

### After
good

### CODE REVIEW

**Comment by iRoachie** [comment]
Hey @saxenanihal95 this doesn't actually fix the problem. Setting `display: none` will still cause other elements on the screen to shift. Display none is the same as not rendering.

**Comment by saxenanihal95** [comment]
@iRoachie my bad i will update the PR

**Comment by saxenanihal95** [comment]
@iRoachie updated PR used opacity instead of display

**Comment by iRoachie** [comment]
@xavier-villelegier We have another problem that will come by using this. Components that use the Input component like the SearchBar will end up displaying with extra space.

image

Maybe instead should add a prop to always show error message ?

**Comment by saxenanihal95** [comment]
I think one way is to separate `InputComponent` which can be used by `Input` and others components like `SearchBar`.

**Comment by iRoachie** [comment]
What I think we can do is add a prop on the input maybe called `renderErrorMessage` that the SearchBar component could use to achieve the styling we have atm.

This will be a breaking change anyway since it affects layout. Let’s do this for 2.0.0

**Comment by saxenanihal95** [comment]
ok @iRoachie so we will add a prop `renderErrorMessage` which will be `true` by default and when `Search` uses Input it will make `renderErrorMessage`  `false` ?

**Comment by iRoachie** [comment]
Yea that's what I was thinking

**Comment by saxenanihal95** [comment]
ok then will update the PR according to that

**Comment by iRoachie** [comment]
Also you need to rebase your branch from `next`. Some conflicts there

**Comment by saxenanihal95** [comment]
ok i will rebase

**Comment by saxenanihal95** [comment]
I have fixed the conflicts now will make changes according to the discussion we have

**Comment by iRoachie** [comment]
There's no changes now in the PR?

**Comment by saxenanihal95** [comment]
yes not right now

**Comment by saxenanihal95** [comment]
@iRoachie Updated the PR

**Comment by iRoachie** [comment]
The default would be true, also you need to add to the typescript definitions.

Also the search bars would need to be updated to set this property to false.

Think we should leave this to 2.0.0 cause it’ll break some layouts that people have in their apps

**Comment by saxenanihal95** [comment]
ok then i will wait for 2.0.0

**Comment by iRoachie** [comment]
@saxenanihal95 Are you still interested in making this feature?

**Comment by saxenanihal95** [comment]
@iRoachie yes

File: src/input/Input.js [hunk 1]
```diff
@@ -138,15 +139,15 @@ class Input extends React.Component {
...
+        {renderErrorMessage && (
           <Text
             {...errorProps}
             style={StyleSheet.flatten([
               styles.error(theme),
               errorStyle && errorStyle,
             ])}
           >
-            {errorMessage}
+            {errorMessage || ''}
```
Comment (iRoachie) [inline src/input/Input.js]:
```suggestion
            {errorMessage}
```

Comment (saxenanihal95) [reply src/input/Input.js]: @iRoachie but In all SearchBar renderErrorMessage is already false ?

File: src/searchbar/SearchBar-android.js [hunk 2]
```diff
@@ -142,6 +142,7 @@ class SearchBar extends Component {
             styles.rightIconContainerStyle,
             rightIconContainerStyle,
           ])}
+          renderErrorMessage={false}
```
[the remark this answers is not in the document]
Comment (iRoachie) [reply src/searchbar/SearchBar-android.js]: Sorry I missed that.

[the remark this answers is not in the document]
Comment (iRoachie) [reply src/searchbar/SearchBar-android.js]: Just the docs and types then

[the remark this answers is not in the document]
Comment (saxenanihal95) [reply src/searchbar/SearchBar-android.js]: ok sure

File: docs/input.md [hunk 3]
```diff
@@ -243,6 +243,14 @@ styling for right Icon Component container
 | :-----------------: | :-----: |
 | View style (object) |  none   |
 
+### `renderErrorMessage`
+
+displays errorMessage
```
Comment (iRoachie) [inline docs/input.md]:
```suggestion
If the error message container should be rendered (take up vertical space). If `false`, when showing errorMessage, the layout will shift to add it at that time.
```

File: src/index.d.ts [hunk 4]
```diff
@@ -879,6 +879,12 @@ export interface InputProps extends TextInputProperties {
    *  props to be passed to the React Native Text component used to display the label (optional)
    */
   labelProps?: TextProps;
+
+  /**
+   *  displays error message
+   */
+
```
Comment (iRoachie) [inline src/index.d.ts]:
```suggestion
```

**Comment by iRoachie** [comment]
You need to update the snapshots for the ci to pass. https://react-native-elements.github.io/react-native-elements/docs/testing.html#snapshot-testing

**Comment by saxenanihal95** [comment]
> You need to update the snapshots for the ci to pass. https://react-native-elements.github.io/react-native-elements/docs/testing.html#snapshot-testing

 @iRoachie Done

**Comment by iRoachie** [comment]
Okay, we need to make some last-minute changes. Right now if `renderErrorMessage={false}` and `errorMessage="Test"` nothing shows at all. Which means right now users would be unable to get back the previous behaviour.

I made a table to explain the different states:

| renderErrorMessage | errorMessage | message | take up space (render)
| -- | -- | -- | -- |
| true | true | yes | yes |
| true | false | no | yes |
| false | false | no | no |
| false | true | yes | yes | 

I've managed to get the correct behaviour by making a constant:

```jsx
const hideErrorMessage = !renderErrorMessage && !errorMessage;


<Text
  {...errorProps}
  style={StyleSheet.flatten([
    styles.error(theme),
    errorStyle && errorStyle,
    hideErrorMessage && {
      height: 0,
      margin: 0,
      padding: 0
    },
  ])}
>
  {errorMessage}
</Text>
```

**Comment by saxenanihal95** [comment]
@iRoachie test is getting passed locally but failing here

**Comment by saxenanihal95** [comment]
@iRoachie any update on this ?

**Comment by saxenanihal95** [comment]
@iRoachie I have updated the snapshots and now tests are passing

**Comment by flyingcircle** [comment]
@saxenanihal95 can you squash your changes into 1 or 2 commits?

**Comment by flyingcircle** [comment]
@saxenanihal95 I tested it out, but I don't see it working. Are you able to share an example of this change working?

**Comment by saxenanihal95** [comment]
@flyingcircle are you talking about jest test ?

**Comment by flyingcircle** [comment]
@saxenanihal95 No, I mean I imported your changes into an app and still saw the error causing a change in rendering margin:
Here was my component:
```
<Input
  containerStyle={[styles.inputContainerStyle]}
  placeholder="Shake input"
  ref={ref => (this.shakeInput = ref)}
  rightIcon={
    <Button
      title="Shake"
      onPress={() => {
          this.shakeInput && this.shakeInput.shake()
          if (this.state.shakeError) {
            this.setState({shakeError: ""});
          } else {
              this.setState({shakeError: "Shake me on error !"});
          }
      }}
    />
  }
  errorMessage={this.state.shakeError}
/>

```

app-error

File: src/input/Input.js [hunk 5]
```diff
@@ -172,6 +178,11 @@ Input.propTypes = {
   labelStyle: TextPropTypes.style,
   labelProps: PropTypes.object,
   theme: PropTypes.object,
+  renderErrorMessage: PropTypes.bool,
+};
+
+Input.defaultProps = {
```
Comment (saxenanihal95) [inline src/input/Input.js:184]: @flyingcircle thanks for pointing out actually defaultProps was already defined and my defined default props was overriding with that

Will update the PR

Comment (saxenanihal95) [reply src/input/Input.js:184]: @flyingcircle I have updated the PR now it will work

**Comment by flyingcircle** [comment]
@saxenanihal95 thanks for your work!

**Comment by AzzouDuGhetto** [comment]
What about the TypeScript definition ? It seems like @iRoachie delete it but why ?

**Comment by flyingcircle** [comment]
@AzzouDuGhetto this looks like it was a merging error on my part. Somehow the removal made it into my PR. Should be fixed.

**Comment by AzzouDuGhetto** [comment]
Forgive my lack of knowledge but isn't #2323 remove it once again ? Or the change will be available later ?

**Comment by flyingcircle** [comment]
@AzzouDuGhetto I pushed a fix to the `next` branch. It will be included in 2.0.1.

**Pull request merged**